src/ipc/dma-copy.c

```diff
@@ -74,10 +74,30 @@ int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
 	return size;
 }
 
+int dma_copy_to_host_nowait(struct dma_copy *dc, struct dma_sg_config *host_sg,
+			    int32_t host_offset, void *local_ptr, int32_t size)
+{
+	int ret;
+
+	/* return if DMA channel is not get yet */
+	if (!dc->chan)
+		return -EINVAL;
+
+	/* tell gateway to copy */
+	ret = dma_copy_legacy(dc->chan, size, 0);
+	if (ret < 0)
+		return ret;
+
+	/* bytes copied */
+	return size;
+}
```

Comment:
Would it make sense the have similar wrapping as with `!CONFIG_DMA_GW`
```c
static int
dma_copy_to_host_flags(struct dma_copy *dc, struct dma_sg_config *host_sg,
		       int32_t host_offset, void *local_ptr, int32_t size,
		       uint32_t flags)
{
	int ret;

	/* tell gateway to copy */
	ret = dma_copy_legacy(dc->chan, size, flags);
	if (ret < 0)
		return ret;

	/* bytes copied */
	return size;
}

int dma_copy_to_host(struct dma_copy *dc, struct dma_sg_config *host_sg,
		     int32_t host_offset, void *local_ptr, int32_t size)
{
	return dma_copy_to_host_flags(dc, host_sg, host_offset, local_ptr, size,
				      DMA_COPY_BLOCKING);
}

int dma_copy_to_host_nowait(struct dma_copy *dc, struct dma_sg_config *host_sg,
			    int32_t host_offset, void *local_ptr, int32_t size)
{
	return dma_copy_to_host_flags(dc, host_sg, host_offset, local_ptr, size, 0);
}
```
Btw, do you know why ONE_SHOT is not applicable for DMA_GW?

Reply:
If we ONE_SHOT is good for both DMAs then we could have only the `dma_copy_to_host_flags()` ifdefed.

Reply:
If I am not mistaken, BLOCKING is the only flag intel/hda-dma.c is sensitive to. But I have no idea about the other dma controllers. You tell me, should I dare to drop ONE_SHOT to make the code tiny bit cleaner? When I last did that I broke the probes :laughing: . Then again there was no non blocking version for !CONFIG_DMA_GW despite the function  being called *_nowait()...